File: anuga_cuda/interpolate_from_vertices_to_edges/interpolate_from_vertices_to_edges.py

import numpy
from pycuda import driver as drv
from anuga_cuda.merimbula_data.channel1 import generate_domain
from anuga_cuda.merimbula_data.generate_domain import domain_create

#domain1 = generate_domain( gpu=False )
#domain2 = generate_domain( gpu=True )
domain1 = domain_create( gpu=False )
domain2 = domain_create( gpu=True )

domain1.protect_against_infinitesimal_and_negative_heights()
domain2.protect_against_infinitesimal_and_negative_heights()

if domain1.optimised_gradient_limiter:
    if domain1._order_ == 1:
        for name in domain1.conserved_quantities:
            Q1 = domain1.quantities[name]
            Q1.extrapolate_first_order()
            
            Q2 = domain2.quantities[name]
            Q2.extrapolate_first_order()
    
    
    elif domain1._order_ == 2:
        domain1.extrapolate_second_order_sw()
        domain2.extrapolate_second_order_sw()
else:
    for name in domain1.conserved_quantities:
        Q1 = domain1.quantities[name]
        Q2 = domain1.quantities[name]
        if domain1._order_ == 1:
            Q1.extrapolate_first_order()
            Q2.extrapolate_first_order()
        elif domain1._order_ == 2:
            Q1.extrapolate_second_order_and_limit_by_vertex()
            Q2.extrapolate_second_order_and_limit_by_vertex()
        
N = domain1.quantities['stage'].vertex_values.shape[0]
W1 = 32
W2 = 1
W3 = 1

domain1.balance_deep_and_shallow()
domain2.balance_deep_and_shallow()

for name in domain1.conserved_quantities:
    Q1 = domain1.quantities[name]
    Q1.interpolate_from_vertices_to_edges()

    Q2 = domain2.quantities[name]
    domain2.interpolate_from_vertices_to_edges_func(
        numpy.int32( N ),
        drv.In( Q2.vertex_values ),
        drv.InOut( Q2.edge_values ),
        block = (W1, W2, W3),
        grid = ( (N + W1*W2*W3 -1) / (W1*W2*W3), 1)
        )

    print name
    print Q1.edge_values, Q2.edge_values
    print numpy.allclose(Q1.edge_values, Q2.edge_values)

File: anuga_cuda/openHMPP/new/interpolate_from_vertices_to_edges.c
//#define REARRANGED_DOMAIN

void interpolate_from_vertices_to_edges(
        int N,
        double* vertex_values,
        double* edge_values) 
{
    int k;

#ifndef REARRANGED_DOMAIN
    int k3;
#endif
    double q0, q1, q2;

    for (k=0; k<N; k++) {

#ifndef REARRANGED_DOMAIN
        k3 = k*3;
        q0 = vertex_values[k3 + 0];
        q1 = vertex_values[k3 + 1];
        q2 = vertex_values[k3 + 2];

        edge_values[k3 + 0] = 0.5*(q1+q2);
        edge_values[k3 + 1] = 0.5*(q0+q2);
        edge_values[k3 + 2] = 0.5*(q0+q1);
#else
        q0 = vertex_values[k];
        q1 = vertex_values[k + N];
        q2 = vertex_values[k + 2*N];

        edge_values[k] = 0.5*(q1+q2);
        edge_values[k + N] = 0.5*(q0+q2);
        edge_values[k + 2*N] = 0.5*(q0+q1);
#endif
    }
}